### ISSUE TITLE
Running parallel instances 

### ISSUE THREAD

**Issue by catubc** [CannyLab/tsne-cuda#33]
Hi. I am wondering if it is possible to make multiple calls to tsnecuda() from a python multiprocessing module?  It would be really great if that could work, but I am guessing the current implementation maxes out threads for each run.

Currently, when I try to call tsnecuda() from multiple CPU threads I get this kind of error:

```
[I 19:42:01.068 NotebookApp] Kernel interrupted: 37453a58-efbd-4902-a13d-921186587399
E: Device warp size not supported.E: Device warp size not supported.

E: Device warp size not supported.
E: Device warp size not supported.
E: Device warp size not supported.
E: Device warp size not supported.
E: Device warp size not supported.

```

Thanks for the help.

**Comment by DavidMChan** [comment]
Hmm, the library wasn't really designed with this in mind, however I can't think of any reason why it wouldn't be supported in our code (officially). It might be that one of our upstream dependencies does something interesting with CUDA contexts and/or shared memory which prevents multiple instances from functioning simultaneously. It's an odd error to get however, since the warp size should be independent of the number of running tsnecuda instances.

**Comment by rmrao** [comment]
Very odd - some minor experimentation suggests that this isn't a problem with *parallel* instances, but rather just with running tsnecuda in a child process (even if there's only 1 process). E.g. the following fails:

```python
import tsnecuda
from multiprocessing import Process

p = Process(target=tsnecuda.test)
p.start()

E: Device warp size not supported
```

I have no idea why this should be the case, unless there's something weird about accessing CUDA from a child process?

**Comment by DavidMChan** [comment]
I suspect that we are leaking some variables in the memory space of the parent process to the memory space of the child process which is causing issues. This may be fixed by #39? @rmrao, can you confirm that #39 fixes this problem?

**Comment by rmrao** [comment]
No it does not - I got the issue with the most recent updates.

-Roshan

**Comment by DavidMChan** [comment]
Hmm. Interesting. This means that it's not in the actual tsne computation, but in the KNN computation or the options/configuration.

**Comment by DavidMChan** [comment]
This is resolved in the 3.0.0 release!
